src/assign.c

```diff
@@ -520,6 +520,8 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
     else if (TRUELENGTH(names) != oldtncol)
       // Use (long long) to cast R_xlen_t to a fixed type to robustly avoid -Wformat compiler warnings, see #5768, PRId64 didn't work
       internal_error(__func__, "selfrefnames is ok but tl names [%lld] != tl [%d]", (long long)TRUELENGTH(names), oldtncol);  // # nocov
+    if (!selfrefok(dt, verbose)) // #6410 setDT(dt) and subsequent attr<- can lead to invalid selfref
+      error(_("It appears that at some ealier point, attributes of this data.table have been reassigned. Please ensure to use setattr() rather than attr<-. Otherwise, please report to data.table issue tracker.")); // # nocov
```

Comment:
```suggestion
      error(_("It appears that at some earlier point, attributes of this data.table have been reassigned. Please use setattr(DT, name, value) rather than attr(DT, name) <- value. If that doesn't apply to you, please report your case to the data.table issue tracker.")); // # nocov
```

Reply:
Should we list a specific way to cause the problem ("use `setattr` instead of `attr<-`") or a more cryptic mechanism of how the problem is caused ("It appears that at some earlier point, this data.table has been shallow-copied")? I think it's `shallow_duplicate` that requires catching the wrong `TRUELENGTH` afterwards:

```
R -q -s -e '
 mtcars <- data.table::setDT(data.table::copy(datasets::mtcars));
 x <- rlang::duplicate(mtcars, FALSE);
 data.table::set(x, NULL, "test", NA_character_)
'
# Error in data.table::set(x, NULL, "test", NA_character_) : 
#   It appears that at some earlier point, names of this data.table have been reassigned. Please ensure to use setnames() rather than names<- or colnames<-. Otherwise, please report to data.table issue tracker.
# Execution halted
R -q -s -e '
 mtcars <- data.table::setDT(data.table::copy(datasets::mtcars));
 x <- rlang::duplicate(mtcars, TRUE);
 data.table::set(x, NULL, "test", NA_character_)
'
# 
#  *** caught segfault ***
# address 0x60000006e, cause 'memory not mapped'
# 
# Traceback:
#  1: data.table::set(x, NULL, "test", NA_character_)
# An irrecoverable exception occurred. R is aborting now ...
```

Reply:
I think the error message remains unclear. Currently, the error message seems to suggest that the user's mistake occurred "at some earlier point". Ideally we should catch it there. I think 'earlier' is redundant and may actually cause confusion.

Is it possible to create a `attr<-.data.table` method which just raises an error. Or is the issue more broad than that.

Reply:
`attr<-` is a primitive function. Unlike [`@<-`](https://github.com/r-devel/r-svn/blob/dff9236d24ca02e4d6f87185bbb2b4e28a9c31cf/src/main/attrib.c#L1617-L1619), it [does not dispatch](https://github.com/r-devel/r-svn/blob/dff9236d24ca02e4d6f87185bbb2b4e28a9c31cf/src/main/attrib.c#L1634-L1664) to S3 methods.

The issue could be more broad than `attr<-` because it is caused by the `data.table` being subject to `shallow_duplicate`. The example above uses `rlang` to cause the issue manually. Don't know whether other, more normal R functions could call `shallow_duplicate` in a similar manner.

Reply:
> I think 'earlier' is redundant and may actually cause confusion.

That wording is shared with a similar error:

https://github.com/Rdatatable/data.table/blob/038ce2b79cc0bbf90cc9f7012dcac82029cad100/src/assign.c#L526

So I don't think we should block the PR based on that. We can file a follow-up issue to improve clarity IMO.

> attr<- is a primitive function. 

Somehow I missed this thread 🙃 https://github.com/Rdatatable/data.table/pull/6419#issuecomment-2316684212